### PR TITLE
client: Add support for NO_PROXY/no_proxy

### LINKED ISSUE #1203: Support for NO_PROXY
### Would you like to work on this feature?

yes

### What problem are you trying to solve?

kube-rs currently supports the HTTP_PROXY (and all its variants) environment variable, but not the NO_PROXY var.

https://github.com/kube-rs/kube/blob/be15f7aab94863b24b7c903dc08812044504b1e3/kube-client/src/config/file_loader.rs#L110-L127

The NO_PROXY var is [not standardised](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/), but it's supported by [Go](https://github.com/golang/go/blob/682a1d2176b02337460aeede0ff9e49429525195/src/vendor/golang.org/x/net/http/httpproxy/proxy.go#L82-L97) and I think we could follow with similar support in here.

### Describe the solution you'd like

change the `proxy_url` fn so that if NO_PROXY is set and the content matches the `cluster.server` url, return `None`, otherwise continue as usual.

The matching impl could be similar to https://github.com/seanmonstar/reqwest/pull/877

### Describe alternatives you've considered

None

### Documentation, Adoption, Migration Strategy

N/A

### Target crate for feature

kube-client

### PR DESCRIPTION
**Pull request by Techassi**
## Motivation

There currently is no support for `NO_PROXY`/`no_proxy` in `kube-client`, see https://github.com/kube-rs/kube/issues/1203. A couple of days ago, a user of our `stackablectl` command line tool reported that their proxy setup does not work as expected. This is due the missing `NO_PROXY` support.

## Solution

Implement `NO_PROXY`/`no_proxy` in `kube-client` by adding a new dedicated module. The implementation currently allows IP addresses, IP ranges (CIDR notation), and plain host names.

There are also a few very simple (positive only) unit tests.

## Notes

A few notes on this PR:

- I left two comments where we could possibly relax the parsing/validation. Let me know your stance on this.
- I'm happy to add more unit tests (also negative ones) with the help of `rstest` for example.

### CODE REVIEW

**Comment by Techassi** [comment]
Okay clippy is complaining about the size of the error enum. I assume it is because I added a new variant to `KubeconfigError`. Let me know, if and how I should deal with it.

File: kube-client/src/config/no_proxy.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+            matchers.push(matcher);
+        }
+
+        Ok(Self::Matchers(matchers))
+    }
+}
+
+impl NoProxy {
+    /// Parse the configuration from the `NO_PROXY` and `no_proxy` environment variables.
+    /// `no_proxy` takes precedence.
```
Comment (Techassi) [inline kube-client/src/config/no_proxy.rs:67]: I actually just noticed that this is not the case. I'm happy to switch the order of env vars to match this, or to change the comment.

Let me know what you think.

Comment (clux) [reply kube-client/src/config/no_proxy.rs:67]: It probably doesn't matter much, but I'd lean to taking the capitalised one as precedence from standard Linux conventions.

As long as the docs match the code.

File: kube-client/Cargo.toml [hunk 2]
```diff
@@ -54,6 +54,7 @@ http-body-util = { workspace = true, optional = true }
 either = { workspace = true, optional = true }
 thiserror.workspace = true
 futures = { workspace = true, optional = true, features = ["std"] }
+ipnet.workspace = true
```
Comment (clux) [inline kube-client/Cargo.toml:57]: I think this dep should be gated behind a proxy feature technically since it is not needed in cluster afaict 

Awkwardly the socks feature and http proxy have their own features though. Maybe it's worth making a general "proxy" feature for all of them?

**Review by clux — COMMENTED** [review]
Thanks a lot! I think this is a nice addition that is l sits nicely orthogonally to other code. Some more unit tests for negative cases (that you have error variants for) would be nice also if possible. 

I have one potentially difficult comment there on features, but we can always raise a follow-up issue for consolidating proxy feature flags (if that is a sensible approach).

File: kube-client/src/config/no_proxy.rs (hunk 1 above)
Comment (Techassi) [reply kube-client/src/config/no_proxy.rs]: Alright, I will adjust the doc comment accordingly then.

File: kube-client/Cargo.toml (hunk 2 above)
Comment (Techassi) [reply kube-client/Cargo.toml]: Okay two things to do here then:

- I'm happy to gate the `ipnet` dependency behind the `http-proxy` feature. That makes sense and I actually wanted to do that in the first place, but seems like I forgot to do it in the end.
- I'm also happy to add a "wrapper" `proxy` feature which enables both `http-proxy` and `socks5`.

**Comment by Techassi** [comment]
> Some more unit tests for negative cases (that you have error variants for) would be nice also if possible.

Happy to add more (negative) unit tests. Any objections introducing `rstest` for making writing these kind of tests easier? I would look like this:

```rust
#[cfg(test)]
mod tests {
    use super::*;

    const CIDR: &str = "10.0.0.0/8";

    #[rstest]
    #[case("10.100.100.100")]
    #[case("10.10.10.10")]
    #[case("10.0.0.1")]
    fn cidr_match(#[case] target: &str) {
        let no_proxy = NoProxy::from_str(CIDR).expect("static input must parse");
        let target_uri = http::Uri::from_str(input).expect("static input must parse");
        assert!(no_proxy.matches(&target_uri).expect("target uri is valid"));
    }

    #[rstest]
    #[case("192.168.0.1")]
    #[case("127.0.0.1")]
    fn cidr_no_match(#[case] target: &str) {
        let no_proxy = NoProxy::from_str(CIDR).expect("static input must parse");
        let target_uri = http::Uri::from_str(input).expect("static input must parse");
        assert!(!no_proxy.matches(&target_uri).expect("target uri is valid"));
    }
}
```

File: kube-client/src/config/no_proxy.rs (hunk 1 above)
Comment (Techassi) [reply kube-client/src/config/no_proxy.rs]: Adjusted in 92587a07.

File: kube-client/Cargo.toml (hunk 2 above)
Comment (Techassi) [reply kube-client/Cargo.toml]: The wrapper `proxy` feature was added in 0b97fd0b and 221e3eaf gates the `ipnet` dependency and dependent code behind any of `socks5` and `http-proxy`. Let me know if that is what you had in mind as I'm not quite sure this is the best way we can do this.

**Comment by clux** [comment]
> Happy to add more (negative) unit tests. Any objections introducing `rstest` for making writing these kind of tests easier?

Sounds good to me. Seems lightweight and well maintained.

File: kube-client/Cargo.toml (hunk 2 above)
Comment (clux) [reply kube-client/Cargo.toml]: This looks very sensible. Thank you. Happy to leave it as this for now!

**Comment by clux** [comment]
> Okay clippy is complaining about the size of the error enum. I assume it is because I added a new variant to `KubeconfigError`. Let me know, if and how I should deal with it.

Ugh. Maybe the biggest error in the can be boxed or something. Yours could probably be boxed as well - but it's probably not the biggest offender. I wouldn't judge you if you slapped a clippy ignore on it and raised a quick (2 sentence) issue with it linking back here. I can look at this in a week.

File: kube-client/Cargo.toml (hunk 2 above)
Comment (Techassi) [reply kube-client/Cargo.toml]: Okay, good. Thank you. I will now mark this as resolved.

**Comment by Techassi** [comment]
> Sounds good to me. Seems lightweight and well maintained.

Okay cool. I will then add a few more unit tests (both positive and negative).

> Yours could probably be boxed as well - but it's probably not the biggest offender.

Yeah I think _that_ one should be fine as it only contains three variants without much data.

> I wouldn't judge you if you slapped a clippy ignore on it and raised a quick (2 sentence) issue with it linking back here. I can look at this in a week.

I'm happy to do that. But before doing that, I will check out the Clippy error in detail and see if it is easy enough to fix.

**Comment by Techassi** [comment]
Status update:

- I added a whole bunch of positive and negative unit tests for `NoProxy` in 0767ac7f.
- I tried `Box`ing the `KubeconfigError` in the `kube_client::Error`, but that still threw Clippy errors. I'm tempted to say we should ignore these errors for now (and allow the lint). I'm happy to tackle this is a separate, proper PR.
  I don't know how far along we are with https://github.com/kube-rs/kube/issues/688, but `Box`ing is the way forward. https://github.com/kube-rs/kube/issues/688 mentions the initial move away from `snafu`, but notes that eventually `snafu` should be used again. This would really help, because with `snafu`, you can do something like:

```rust
enum Error {
    #[snafu(display("failed to get resource"))]
    GetResource {
        #[snafu(source(from(crate::client::Error, Box::new)))] // <<< See here
        source: Box<crate::client::Error>,
    }
}
```

**Comment by clux** [comment]
All your changes look great.

The refine error issue is quite old and not sure how accurate it is these days. No one has had a lot of energy to deal with error handling choices. There are opinions going both ways. I feel we should probably go with ignoring the error for now. I might box up a few later to keep the number lower. Will have more time to look at this properly next week. Happy to merge once clippy is silenced.

**Comment by Techassi** [comment]
~~Alright, I will add the appropriate `allow`s then!~~

I added `kube_runtime::watcher::Error` to the "large error" ignore list as per Clippy's [`result_large_err`](https://rust-lang.github.io/rust-clippy/main/index.html#result_large_err)'s config.

**Comment by clux** [comment]
LGTM. Thank you! Waiting for github being slow today, then merging if green 😄

**Comment by Techassi** [comment]
I was noticing it as well... Super slow...

Thank you for merging in advance (if green)!

**Comment by Techassi** [comment]
Ah well, I guess I have to add more ignores for:

- `kube_runtime::finalizer::Error`
- `kube_runtime::wait::delete::Error`
- `kube_runtime::wait::Error`

**Comment by Techassi** [comment]
I think the integration test was flaky. As far as I can see it succeeded multiple times before.

**Comment by doxxx93** [comment]
Found the cause, and it's on our side rather than the test. Event names are `{name}.{nanos:x}` from `Timestamp::now()` (kube-runtime/src/events.rs:274), and both failures we've had (this one, and `event_recorder_attaches_events` on 2026-09-14 in run 34830841340) carry sub-microsecond digits, so two creates landing on the same nanosecond isn't a plausible explanation. The same POST went out twice.

`RetryPolicy::retry` decides on the response status alone and ignores the request it was handed (`_req`, kube-client/src/client/retry.rs:140), so any method is replayed on 429/503/504 with the body cloned, and that policy has been in the default client stack since #2007. A create is not idempotent, so replaying one the apiserver already persisted comes back as 409 AlreadyExists. I reproduced it against a fake apiserver: answer the first POST with 503 and the client re-sends an identical `metadata.name`.

Opening an issue with the repro and a fix. Nothing for this PR to do, the rerun was the right call.

**Comment by Techassi** [comment]
Oh wow, what an edge-case. Thank you for looking into this so thoroughly!

**Review by clux — APPROVED** [review]

**Comment by Techassi** [comment]
Thank you very much for merging this!

I will work on a bunch of follow-up tasks which this PR produced:

- Adding even more unit tests for this. Done in [#2083](https://github.com/kube-rs/kube/pull/2083)
- Making local and CI checks more consistent
- Maybe work on the error size issues
